File: fibonacci.exs
defmodule Fibonacci do
  def fibonacci_hof(iterator) do
    fn (repeat) ->
      iterate_fibonacci(repeat, iterator)
    end
  end

  defp iterate_fibonacci(0, _), do: :ok
  defp iterate_fibonacci(n, iterator) do
    term = fibonacci(n)
    iterator.(term)
    iterate_fibonacci(n - 1, iterator)
  end

  def fibonacci(0), do: 0
  def fibonacci(1), do: 1
  def fibonacci(n), do: fibonacci(n - 1) + fibonacci(n - 2)
end

defmodule Main do
  def print_next_term(s) do  
    IO.puts("Fibonacci HOF: #{s}")
  end
end

fibonacci_with_print = Fibonacci.fibonacci_hof(&Main.print_next_term/1)

fibonacci_with_print.(10)

File: fibonacci.js


const d1 = document.getElementById("result")

const d2 = document.getElementById("resultHFO")

d1.innerHTML = fibonacci(5)

function fibonacci(repeat){
  let n1 = 0, n2 = 1, next;
  
  next = n1 + n2;

  for (let i = 0; i < repeat; i++) {


    n1 = n2;
    n2 = next;
    next = n1 + n2;
    console.log("Fibonacci: " + next);
}
  return next
}

// ------------------------------------------- HOF ---------------------------------- //
const fibonacciHOF = (iterator) => {
    return (repeat) => {
    let n1 = 0, n2 = 1, next;

    next = n1 + n2;

    for (let i= 0; i < repeat; i++) {
        n1 = n2;
        n2 = next;
        next = n1 + n2;
        iterator(next);
    }
    return next;
    }
}

function printNextTerm(s) {
    console.log("Fibonacci HOF: " + s);
        
}

const fibonacciWithPrint = fibonacciHOF(printNextTerm);

// ------------------------------------------- HOF ---------------------------------- //

d2.innerHTML = fibonacciWithPrint(4)

